File: cpw2017/games/frogger.py
"""
Example game that you can use as a template for your own games!

This code contains a lot of useful functions that you can use
for coming up with ideas for your own games, such as how you
move characters around.
"""

import sys
import time
import pygame
from pygame.locals import *

# Window dimensions
WINDOW_WIDTH = 550
WINDOW_HEIGHT = 400

# Game board color
#
# NOTE: The only color the board
# buttons can turn is red.
RED = 255, 0, 0
BLACK = 0, 0, 0
FILL_COLOR = RED

# Key syntactic sugars
KEY_W = 0
KEY_A = 1
KEY_S = 2
KEY_D = 3

KEY_UP = 4
KEY_DOWN = 5
KEY_LEFT = 6
KEY_RIGHT = 7

# Game global variables
KEYS = None
CHAR_X = None
CHAR_Y = None
SURFACE = None
TIMEOUT = None
GAME_COUNTER = None
SCREEN_BUFFER = None


def main():
    global CHAR_X, CHAR_Y, GAME_COUNTER, KEYS, SCREEN_BUFFER, SURFACE, TIMEOUT

    pygame.init()

    title = "Pygame Keyboard Test"
    SURFACE = pygame.display.set_mode((WINDOW_WIDTH, WINDOW_HEIGHT))

    pygame.display.set_caption(title)
    pygame.mouse.set_visible(False)

    # The display counter used for setting how often
    # we update the screen buffer.
    GAME_COUNTER = 0

    # The screen is a 7x5 array and is the only item
    # that should be modified during your game.
    SCREEN_BUFFER = [[0, 0, 0, 0, 0],
                     [0, 0, 0, 0, 0],
                     [0, 0, 0, 0, 0],
                     [0, 0, 0, 0, 0],
                     [0, 0, 0, 0, 0],
                     [0, 0, 0, 0, 0],
                     [0, 0, 0, 0, 0]]

    # Location for a character pixel
    CHAR_X = 0
    CHAR_Y = 0

    # Counter used when moving a character
    TIMEOUT = 0

    # Array for keeping tracked of relevant
    # game keys and whether they have been
    # pressed or not.
    #
    # 0 = Key Not Pressed
    # 1 = Key Pressed
    KEYS = [0, 0, 0, 0, 0, 0, 0, 0]

    while True:
        # key_test()
        # move_character()
        frogger()

        # Needed for simulation to run.
        update_simulator()


def clear_buffer():
    """
    Clear the screen buffer.
    """

    for x in range(7):
        for y in range(5):
            SCREEN_BUFFER[x][y] = 0


def draw_screen():
    """
    Redraw the screen using the updated screen buffer.
    """

    for x in range(7):
        for y in range(5):
            radius = 30
            pos = 50 + 75*x, 350 - 75*y

            if SCREEN_BUFFER[x][y] == 0:  # unfilled
                pygame.draw.circle(SURFACE, FILL_COLOR, pos, radius, 3)
            else:  # filled
                pygame.draw.circle(SURFACE, FILL_COLOR, pos, radius, 0)


def update_keys(new_keys):
    """
    Update the key mappings based on the new keys provided.

    :param new_keys: The mappings provided from pygame for keys pressed.
    """

    KEYS[KEY_UP] = new_keys[pygame.K_UP]
    KEYS[KEY_DOWN] = new_keys[pygame.K_DOWN]
    KEYS[KEY_RIGHT] = new_keys[pygame.K_RIGHT]
    KEYS[KEY_LEFT] = new_keys[pygame.K_LEFT]

    KEYS[KEY_W] = new_keys[pygame.K_w]
    KEYS[KEY_S] = new_keys[pygame.K_s]
    KEYS[KEY_D] = new_keys[pygame.K_d]
    KEYS[KEY_A] = new_keys[pygame.K_a]


def update_simulator():
    """
    Update the simulator, including the display and keys pressed.
    """

    global GAME_COUNTER

    for event in pygame.event.get():
        if event.type == QUIT:
            sys.exit()
        elif event.type == KEYDOWN and event.key == K_ESCAPE:
            pygame.display.quit()
            sys.exit(0)

    # Update key presses by checking
    # for pressed keys.
    keys = pygame.key.get_pressed()
    update_keys(keys)

    if GAME_COUNTER >= 20:
        draw_screen()
        pygame.display.flip()
        GAME_COUNTER = 0
    else:
        GAME_COUNTER += 1

    SURFACE.fill(BLACK)
    time.sleep(0.001)


# Define test programs here.
def key_test():
    """
    Test program to make sure that we're responding to all relevant keys.
    """

    clear_buffer()

    SCREEN_BUFFER[5][2] = KEYS[KEY_UP]
    SCREEN_BUFFER[5][0] = KEYS[KEY_DOWN]
    SCREEN_BUFFER[6][1] = KEYS[KEY_RIGHT]
    SCREEN_BUFFER[4][1] = KEYS[KEY_LEFT]

    SCREEN_BUFFER[1][2] = KEYS[KEY_W]
    SCREEN_BUFFER[1][0] = KEYS[KEY_S]
    SCREEN_BUFFER[2][1] = KEYS[KEY_D]
    SCREEN_BUFFER[0][1] = KEYS[KEY_A]


def move_character():
    """
    Move a character on the game board.
    """

    global FROG, TIMEOUT

    if TIMEOUT >= 100:
        clear_buffer()
        if KEYS[KEY_UP]:
            FROG[1] += 1
            TIMEOUT = 0
        if KEYS[KEY_LEFT]:
            FROG[0] -= 1
            TIMEOUT = 0
        if KEYS[KEY_DOWN]:
            FROG[1] -= 1
            TIMEOUT = 0
        if KEYS[KEY_RIGHT]:
            FROG[0] += 1
            TIMEOUT = 0
    else:
        TIMEOUT += 1

    # Check bounds on our new location.
    if FROG[0] < 0:
        FROG[0] = 0
    if FROG[0] > 6:
        FROG[0] = 6
    if FROG[1] < 0:
        FROG[1] = 0
    if FROG[1] > 4:
        FROG[1] = 4

    # Output character dot.
    SCREEN_BUFFER[FROG[0]][FROG[1]] = 2


################################
# WRITE YOUR OWN PROGRAM HERE! #
################################

# Note: your code will run every 1 ms, and no other timing mechanisms are
# allowed, so you cannot use time.sleep(). Instead, make a counter variable
# and count time in milliseconds

# Define game global variables here. 

# Frog position [column, row]
FROG = [3, 1]

# Within each row, each truck has the same speed and the same length.
# The trucks list is a list of lists where each sublist represents a row.
# Within each sublist, the first element denotes the length of the trucks in that row (2 to 4)
# The second element denotes the velocity (-2, -1, 1, or 2) with positive being to the right
# Subsequent elements denote the starting position of each truck
PRIOR_LEVEL = []
CURRENT_LEVEL = []
NEXT_LEVEL = []
START_ROW = 0

TRUCK_VELOCITIES = [-1,2,1,-2]
TRUCK_VELOCITY_COUNTER = 0

TRUCK_LENGTH = [2,3,4]
TRUCK_LENGTH_COUNTER = 0

# A blink counter (BLINK) and its reset point (BLINK_RATE)
BLINK = 0
BLINK_RATE = 70

# Truck movement counter (TRUCK_COUNTER) and its reset point (TRUCK_RATE)
TRUCK_COUNTER = 0
TRUCK_RATE = 400
# Variable to make truck move on every other cycle
TRUCK_HALF = 0

# Define program here.
def frogger():
    global TIMEOUT, FROG, BLINK, BLINK_RATE, TRUCK_COUNTER, CURRENT_LEVEL, PRIOR_LEVEL, NEXT_LEVEL

    CURRENT_LEVEL = generate_level()
    PRIOR_LEVEL = generate_level()
    NEXT_LEVEL = generate_level()

    if TIMEOUT >= 50:
        update_screen()
        update_frog()
        update_trucks()
    else:
        TIMEOUT += 1
        BLINK += 1
        TRUCK_COUNTER += 1

def generate_level():
    global TRUCK_VELOCITY_COUNTER, TRUCK_LENGTH_COUNTER
    level = [[0,0],[0,0],[0,0]]
    for row in range(7):
        new_row = [TRUCK_LENGTH[TRUCK_LENGTH_COUNTER], TRUCK_VELOCITIES[TRUCK_VELOCITY_COUNTER]]
        start_position = 1
        while start_position < 7:
            new_row.append(start_position)
            start_position += new_row[0] + abs(new_row[1]) + 1
        TRUCK_LENGTH_COUNTER = (TRUCK_LENGTH_COUNTER + 1) % len(TRUCK_LENGTH)
        TRUCK_VELOCITY_COUNTER = (TRUCK_VELOCITY_COUNTER + 1) % len(TRUCK_VELOCITIES)
        level.append(new_row)
    return level

def update_screen():
    clear_buffer()

def update_frog():
    global FROG, BLINK, BLINK_RATE, SCREEN_BUFFER, TIMEOUT

    key_press = False

    if KEYS[KEY_UP] and not(FROG[1] >= 4):
        FROG[1] += 1
        key_press = True
        TIMEOUT = 0
    if KEYS[KEY_LEFT] and not(FROG[0] <= 0):
        FROG[0] -= 1
        key_press = True
        TIMEOUT = 0
    if KEYS[KEY_DOWN] and not(FROG[1] <= 0):
        FROG[1] -= 1
        key_press = True
        TIMEOUT = 0
    if KEYS[KEY_RIGHT] and not(FROG[0] >= 6):
        FROG[0] += 1
        key_press = True
        TIMEOUT = 0

    #TODO: fix buggy response time with blinking
    if BLINK <= BLINK_RATE:
        SCREEN_BUFFER[FROG[0]][FROG[1]] = 2
    elif BLINK <= 2 * BLINK_RATE:
        SCREEN_BUFFER[FROG[0]][FROG[1]] = 0
    elif BLINK >= 2 * BLINK_RATE:
        BLINK = 0

    #TODO: add collision detection
    
def update_trucks():
    """
    Update position of trucks
    """
    global CURRENT_LEVEL, SCREEN_BUFFER, TIMEOUT, TRUCK_COUNTER, TRUCK_HALF

    if TRUCK_COUNTER >= TRUCK_RATE:
        for row in CURRENT_LEVEL:

            if row[1] == 2 or (row[1] == 1 and TRUCK_HALF == 1):
                for i in range(2, len(row)):
                    row[i] += 1
            elif row[1] == -2 or (row[1] == -1 and TRUCK_HALF == 1):
                for i in range(2,len(row)):
                    row[i] -= 1

            # Post-processing: delete trucks that are gone, add new ones as needed, 
            # Process positive velocities
            if row[1] == 2 or (row[1] == 1 and TRUCK_HALF == 1):

                # Check if the end of the last truck extends beyond the screen
                if row[-1] - row[0] + 1 > 6:
                    del(row[-1])

                # Check if the first truck has advanced far enough to add another truck
                # Sufficient gap is determined by velocity + 1 (row[1])
                if row[2] - row[0] + 1 > row[1]:
                    # Add a truck at position 0
                    row.insert(2,0)

            # Process negative velocities
            elif row[1] == -2 or row[1] == -1 and TRUCK_HALF == 1:
                if row[2] + row[0] -1 < 0:
                    del(row[2])

                if row[-1] + row[0] -1< 7 + row[1]:
                    row.append(7)
        
        # If trucks were updated, reset the truck counter
        TRUCK_COUNTER = 0
        TRUCK_HALF = (TRUCK_HALF + 1) % 2

        # Write all trucks to screen
        for row_num in range(START_ROW, min(START_ROW + 6, 10)):
            for truck in CURRENT_LEVEL[row_num][2:]:
                # Write the pixels for each truck
                for offset in range(CURRENT_LEVEL[row_num][0]):
                    if CURRENT_LEVEL[row_num][1] < 0 and truck + offset < 7 and truck + offset > -1:
                        SCREEN_BUFFER[truck + offset][row_num - START_ROW] = 2
                    elif CURRENT_LEVEL[row_num][1] > 0 and truck - offset > -1 and truck - offset < 7:
                        # print("row_num: {} \ntruck - offset: {}\n".format(row_num, truck - offset))
                        SCREEN_BUFFER[truck - offset][row_num - START_ROW] = 2
                    
    TIMEOUT = 0

def scroll():
    """
    Deletes bottom-most row of TRUCKS that moves off the screen.
    Create new row of trucks and adds it to top of TRUCKS.
    Moves frog position down to maintain its relative position
    """
    global FROG, TRUCKS, SCREEN_BUFFER, TRUCK_LENGTH_COUNTER, TRUCK_VELOCITY_COUNTER

    # Move frog down one position to allow scrolling.
    #if FROG[1] == 5




    FROG[1] -= 1

    del(TRUCKS[0])

    new_row = [TRUCK_LENGTH[TRUCK_LENGTH_COUNTER], TRUCK_VELOCITIES[TRUCK_VELOCITY_COUNTER]]
    TRUCK_VELOCITY_COUNTER = (TRUCK_VELOCITY_COUNTER + 1) % 4
    TRUCK_LENGTH_COUNTER = (TRUCK_LENGTH_COUNTER + 1) % 3

    # Add a little "randomness" to the starting point of the first truck of each new row
    truck_position = TRUCK_LENGTH_COUNTER
    
    # Add trucks to the new row
    while truck_position < 7:
        new_row.append(truck_position)

        # Beginning of next truck must be after the length of the first truck and with a gap
        # Gap is either 2 or 3 as determined by the pairity of TRUCK_VELOCITY_COUNTER
        truck_position += new_row[0] + (TRUCK_VELOCITY_COUNTER % 2 + 3)

    TRUCKS.append(new_row)    


if __name__ == "__main__":
    main()

File: cpw2017/games/frogger.c
/*
 * Copyright (c) 2017, Next Make
 *
 * Move and rotate falling blocks to form lines and clear the board!
 * Game over if the screen fills up.
 *
 * The game speeds up over time, see how long you can last!
 */

#include <avr/io.h>
#define F_CPU 8000000
#include <util/delay.h>

// Threshold for button state
#define BUT 5

// If the LED matrix is in backwards
char correctLED = 0;

// Current LED output:
//
// 0 = off
// 1 = dim
// 2 = bright
unsigned char outputLED[7][5];
char buttons[8];

unsigned char row = 0;
unsigned char everyNth = 0;
unsigned char everyFourth = 0;

/**
 * Turn on/off the appropriate LEDs using the outputLED array.
 */
void writeLeds() {
    if (correctLED) {
        // turn off all rows
        DDRC |= (1 << PC1);
        DDRC |= (1 << PC2);
        DDRC |= (1 << PC3);
        DDRC |= (1 << PC4);
        DDRC |= (1 << PC5);

        PORTC |= (1 << PC1);
        PORTC |= (1 << PC2);
        PORTC |= (1 << PC3);
        PORTC |= (1 << PC4);
        PORTC |= (1 << PC5);

        // Turn on correct row
        if (row == 1) {
            PORTC &= ~(1 << PC1);
        } else if (row == 2) {
            PORTC &= ~(1 << PC2);
        } else if (row == 3) {
            PORTC &= ~(1 << PC3);
        } else if (row == 4) {
            PORTC &= ~(1 << PC4);
        } else {
            PORTC &= ~(1 << PC5);
        }

        // Turn on LEDs in that row
        for (unsigned char i = 0; i < 7; i++) {
            if (outputLED[i][row] == 0) {
                DDRD |= (1 << (i + 1));
                PORTD &= ~(1 << (i + 1));
            } else if (outputLED[i][row] == 1) {
                DDRD &= ~(1 << (i + 1));
                PORTD |= (1 << (i + 1));
            } else {
                DDRD |= (1 << (i + 1));
                PORTD |= (1 << (i + 1));
            }
        }
    } else {
        // Turn off all rows
        DDRC |= (1 << PC4);
        DDRD |= (1 << PD4);
        DDRD |= (1 << PD5);
        DDRC |= (1 << PC1);
        DDRD |= (1 << PD3);

        PORTC |= (1 << PC4);
        PORTD |= (1 << PD4);
        PORTD |= (1 << PD5);
        PORTC |= (1 << PC1);
        PORTD |= (1 << PD3);

        // Turn off all columns
        DDRD |= (1 << PD7);
        PORTD &= ~(1 << PD7);
        DDRD |= (1 << PD6);
        PORTD &= ~(1 << PD6);
        DDRC |= (1 << PC5);
        PORTC &= ~(1 << PC5);
        DDRC |= (1 << PC2);
        PORTC &= ~(1 << PC2);
        DDRC |= (1 << PC3);
        PORTC &= ~(1 << PC3);
        DDRD |= (1 << PD2);
        PORTD &= ~(1 << PD2);
        DDRD |= (1 << PD1);
        PORTD &= ~(1 << PD1);

        // Turn on correct row
        if (row == 0) {
            PORTC &= ~(1 << PC4);
        } else if (row == 1) {
            PORTD &= ~(1 << PD4);
        } else if (row == 2) {
            PORTD &= ~(1 << PD5);
        } else if (row == 3) {
            PORTC &= ~(1 << PC1);
        } else {
            PORTD &= ~(1 << PD3);
        }

        // Turn on LEDs in that row
        //
        // Column 1
        if (outputLED[6][4 - row] == 0) {
        } else if (outputLED[6][4 - row] == 2) {
            PORTD |= (1 << PD7);
        } else if (everyFourth == 1) {
            PORTD |= (1 << PD7);
        }

        // Column 2
        if (outputLED[5][4 - row] == 0) {
        } else if (outputLED[5][4 - row] == 2) {
            PORTD |= (1 << PD6);
        } else if (everyFourth == 1) {
            PORTD |= (1 << PD6);
        }

        // Column 3
        if (outputLED[4][4 - row] == 0) {
        } else if (outputLED[4][4 - row] == 2) {
            PORTC |= (1 << PC5);
        } else if (everyFourth == 1) {
            PORTC |= (1 << PC5);
        }

        // Column 4
        if (outputLED[3][4 - row] == 0) {
        } else if (outputLED[3][4 - row] == 2) {
            PORTC |= (1 << PC2);
        } else if (everyFourth == 1) {
            PORTC |= (1 << PC2);
        }

        // Column 5
        if (outputLED[2][4 - row] == 0) {
        } else if (outputLED[2][4 - row] == 2) {
            PORTC |= (1 << PC3);
        } else if (everyFourth == 1) {
            PORTC |= (1 << PC3);
        }

        // Column 6
        if (outputLED[1][4 - row] == 0) {
        } else if (outputLED[1][4 - row] == 2) {
            PORTD |= (1 << PD2);
        } else if (everyFourth == 1) {
            PORTD |= (1 << PD2);
        }

        // Column 7
        if (outputLED[0][4 - row] == 0) {
        } else if (outputLED[0][4 - row] == 2) {
            PORTD |= (1 << PD1);
        } else if (everyFourth == 1) {
            PORTD |= (1 << PD1);
        }
    }

    row += 1;

    if (row >= 5) {
        row = 0;
        everyNth += 1;
        if (everyNth == 4) {
            everyFourth = 1;
            everyNth = 0;
        } else {
            everyFourth = 0;
        }
    }
}

/**
 * Clear the outputLED array.
 */
void clearLeds() {
    for (unsigned char i = 0; i < 7; i++) {
        for (unsigned char j = 0; j < 5; j++) {
            outputLED[i][j] = 0;
        }
    }
}

/**
 * Get the buttons pressed on the board.
 */
void getButtons() {
    // Button 0
    if (PINB & (1 << PB0)) {
        buttons[0] -= 1;
    } else {
        buttons[0] += 1;
    }

    // Button 1
    if (PINA & (1 << PA1)) {
        buttons[1] -= 1;
    } else {
        buttons[1] += 1;
    }

    // Button 2
    if (PINB & (1 << PB6)) {
        buttons[2] -= 1;
    } else {
        buttons[2] += 1;
    }

    // Button 3
    if (PINB & (1 << PB1)) {
        buttons[3] -= 1;
    } else {
        buttons[3] += 1;
    }

    // Button 4
    if (PINA & (1 << PA3)) {
        buttons[4] -= 1;
    } else {
        buttons[4] += 1;
    }

    // Button 5
    if (PINA & (1 << PA0)) {
        buttons[5] -= 1;
    } else {
        buttons[5] += 1;
    }
    // Button 6
    if (PINB & (1 << PB2)) {
        buttons[6] -= 1;
    } else {
        buttons[6] += 1;
    }
    // Button 7
    if (PINA & (1 << PA2)) {
        buttons[7] -= 1;
    } else {
        buttons[7] += 1;
    }

    for (unsigned char i = 0; i < 8; i++) {
        if (buttons[i] > 20) {
            buttons[i] = 20;
        } else if (buttons[i] < 0) {
            buttons[i] = 0;
        }
    }
}

/**
 * Frogger game variables.
 */
// Current frog position
unsigned char frogX = 3;
// frogY is always 1, we scroll the screen instead of moving the frog

// Counter used for blinking the frog
unsigned char blink = 0;

// Counter used for moving the frog and trucks
unsigned char timeout = 0;

// The trucks in the level are in froggerLevel:
// Within each row, each truck has the same speed and same length.
// The speed of a row is equal to the number of spaces between trucks.
// Within each sublist, the first element is the length of the trucks in that
// row (2-4).
// The second element is the velocity (-2, -1, 1, or 2) with positive velocity
// meaning moving to the right.
// The third element is the truck index, which indicates the portion of the
// first truck in a row that is still off the screen.
// A sublist with all 0s represents an empty row.
char froggerLevel[37][3] = {{0, 0, 0}, {0, 0, 0}, {0, 0, 0}, {2, -2, 0}, {3, 1, 0}, {2, -1, 0}, {0, 0, 0}, {0, 0, 0}, {0, 0, 0}, {1, 2, 0}, {4, -1, 0}, {1, 2, 0},{0, 0, 0}, {0, 0, 0}, {0, 0, 0}, {2, -2, 0}, {3, 1, 0}, {2, -1, 0}, {0, 0, 0}, {0, 0, 0}, {0, 0, 0}, {1, 2, 0}, {1, -2, 0}, {1, 2, 0}, {0, 0, 0}, {0, 0, 0}, {0, 0, 0}, {2, -2, 0}, {3, 1, 0}, {2, -1, 0}, {0, 0, 0}, {0, 0, 0}, {0, 0, 0}, {1, 2, 0}, {4, -2, 0}, {1, 2, 0}, {0, 0, 0}};

// Counter used for moving trucks
unsigned int truckCounter = 0;

// Variable to make speed 1 trucks move on every other cycle
unsigned char truckPhase = 0;

// The beginning of the window of rows currently displayed on the screen
unsigned char rowIndex = 0;

/**
 * Resets variables and returns to the menu.
 */
void froggerEndGame() {
    rowIndex = 0;
    frogX = 3;
}

/**
 * Check if the frog has collided with a truck.
 */
void froggerCheckCollision() {
    unsigned char currentRow = rowIndex + 1;
    if (currentRow < 37) {
        unsigned char index = froggerLevel[currentRow][2];
        if (froggerLevel[currentRow][1] > 0) {
            // Moving to the right
            for (unsigned char col = 0; col < 7; col++) {
                if (index >= froggerLevel[currentRow][1]) {
                    // A truck exists here
                    if (col == frogX) {
                        // Collision!
                        froggerEndGame();
                    }
                }
                index = (index + 1) % (froggerLevel[currentRow][0] + froggerLevel[currentRow][1]);
            }
        } else if (froggerLevel[currentRow][1] < 0) {
            // Moving to the left
            for (unsigned char col = 0; col < 7; col++) {
                if (index < -froggerLevel[currentRow][1]) {
                    // A truck exists here
                    if (col == frogX) {
                        // Collision!
                        froggerEndGame();
                    }
                }
                index = (index - 1) % (froggerLevel[currentRow][0] - froggerLevel[currentRow][1]);
            }
        }
    }
}

/**
 * Move the frog on the screen.
 * The screen scrolls up and down with the frog to show more of the froggerLevel.
 */
void updateFrog() {
    // Up
    if (buttons[1] > 3) {
        rowIndex += 1;
    }
    // Down
    if (buttons[7] > 3 && rowIndex > 0) {
        rowIndex -= 1;
    }
    // Left
    if (buttons[4] > 3 && frogX > 0) {
        frogX -= 1;
    }
    // Right
    if (buttons[5] > 3 && frogX < 6) {
        frogX += 1;
    }
    timeout = 0;

    froggerCheckCollision();
    if (rowIndex == 37) {
        // Victory!
        froggerEndGame();
    }
}

/**
 * Draw the frog on the screen.
 */
void drawFrog() {
    if (blink <= 70) {
        outputLED[frogX][1] = 2;
    } else if (blink <= 140) {
        outputLED[frogX][1] = 0;
    } else {
        blink = 0;
    }
}

/**
 * Update the position of the trucks on the screen.
 */
void updateTrucks() {
    truckCounter += 1;

    if (truckCounter >= 400) {
        truckCounter = 0;
        truckPhase = (truckPhase + 1) % 2;

        for (unsigned char row = 0; row < 37; row++) {
            // Update truck indices
            if (froggerLevel[row][1] == 2 || ((froggerLevel[row][1] == 1) && truckPhase)) {
                froggerLevel[row][2] = (froggerLevel[row][2] - 1) % (froggerLevel[row][0] + froggerLevel[row][1]);
            } else if (froggerLevel[row][1] == -2 || ((froggerLevel[row][1] == -1) && truckPhase)) {
                froggerLevel[row][2] = (froggerLevel[row][2] - 1) % (froggerLevel[row][0] - froggerLevel[row][1]);
            }
        }

        froggerCheckCollision();
    }
}

/**
 * Draw the trucks on the screen.
 */
void drawTrucks() {
    unsigned char maxRow = rowIndex + 5;
    if (maxRow > 37) {
        maxRow = 37;
    }
    // Draw the current trucks
    for (unsigned char row = rowIndex; row < maxRow; row++) {
        unsigned char index = froggerLevel[row][2];
        if (froggerLevel[row][1] > 0) {
            // Trucks moving to the right
            for (unsigned char col = 0; col < 7; col++) {
                if (index >= froggerLevel[row][1]) {
                    // Drawing a truck
                    outputLED[col][row - rowIndex] = 2;
                }
                index = (index + 1) % (froggerLevel[row][0] + froggerLevel[row][1]);
            }
        } else if (froggerLevel[row][1] < 0) {
            // Trucks moving to the left
            for (unsigned char col = 0; col < 7; col++) {
                if (index < -froggerLevel[row][1]) {
                    // Drawing a truck
                    outputLED[col][row - rowIndex] = 2;
                }
                index = (index - 1) % (froggerLevel[row][0] - froggerLevel[row][1]);
            }
        }
    }
}

/**
 * The main Frogger game!
 */
void frogger() {
    clearLeds();

    drawFrog();
    drawTrucks();
    if (timeout >= 75) {
        updateFrog();
        updateTrucks();
    } else {
        timeout += 1;
        blink += 1;
        truckCounter += 1;
    }
}

int main(void) {
    DDRC = 0b10111111;
    DDRD = 0b11111111;
    PORTC = 0b00000000;
    PORTD = 0b00000000;

    // Set up things for button inputs
    DDRA &= ~(1 << PA0);  // BUT_R_R
    PORTA |= (1 << PA0);
    DDRA &= ~(1 << PA1);  // BUT_R_U
    PORTA |= (1 << PA1);
    DDRA &= ~(1 << PA2);  // BUT_R_D
    PORTA |= (1 << PA2);
    DDRA &= ~(1 << PA3);  // BUT_R_L
    PORTA |= (1 << PA3);

    DDRB &= ~(1 << PB0);  // PB0 BUT_L_U
    PORTB |= (1 << PB0);
    DDRB &= ~(1 << PB1);  // PB1 BUT_L_R
    PORTB |= (1 << PB1);
    DDRB &= ~(1 << PB2);  // PB2 BUT_L_D
    PORTB |= (1 << PB2);
    DDRB &= ~(1 << PB6);  // PB6 BUT_L_L
    PORTB |= (1 << PB6);

    while (1) {
        frogger();
        writeLeds();
        getButtons();

        _delay_ms(1);
    }
}
